nupic/embodied/algos/sparse_weights_sac.py

```diff
@@ -0,0 +1,92 @@
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#  See the GNU Affero Public License for more details.
+#
+#  You should have received a copy of the GNU Affero Public License
+#  along with this program.  If not, see http://www.gnu.org/licenses.
+#
+#  http://numenta.org/licenses/
+#
+# ------------------------------------------------------------------------------
+from garage.torch.algos.sac import SAC
+from nupic.torch.modules.sparse_weights import rezero_weights
+
+
+class SparseWeightsSAC(SAC):
+
+    def optimize_policy(self, samples_data):
+        """Optimize the policy q_functions, and temperature coefficient. Rezero
+        model weights (if applicable) after each optimizer step.
+
+        Args:
+            samples_data (dict): Transitions(S,A,R,S') that are sampled from
+                the replay buffer. It should have the keys 'observation',
+                'action', 'reward', 'terminal', and 'next_observations'.
+
+        Note:
+            samples_data's entries should be torch.Tensor's with the following
+            shapes:
+                observation: :math:`(N, O^*)`
+                action: :math:`(N, A^*)`
+                reward: :math:`(N, 1)`
+                terminal: :math:`(N, 1)`
+                next_observation: :math:`(N, O^*)`
+
+        Returns:
+            torch.Tensor: loss from actor/policy network after optimization.
+            torch.Tensor: loss from 1st q-function after optimization.
+            torch.Tensor: loss from 2nd q-function after optimization.
+
+        """
+        obs = samples_data["observation"]
+        qf1_loss, qf2_loss = self._critic_objective(samples_data)
+
+        self._qf1_optimizer.zero_grad()
+        qf1_loss.backward()
+        self._qf1_optimizer.step()
+        self._qf1.apply(rezero_weights)
+
+        self._qf2_optimizer.zero_grad()
+        qf2_loss.backward()
+        self._qf2_optimizer.step()
+        self._qf2.apply(rezero_weights)
+
+        action_dists = self.policy(obs)[0]
+        new_actions_pre_tanh, new_actions = (
+            action_dists.rsample_with_pre_tanh_value())
+        log_pi_new_actions = action_dists.log_prob(
+            value=new_actions, pre_tanh_value=new_actions_pre_tanh)
+
+        policy_loss = self._actor_objective(samples_data, new_actions,
+                                            log_pi_new_actions)
+        self._policy_optimizer.zero_grad()
+        policy_loss.backward()
+
+        self._policy_optimizer.step()
+        self.policy.apply(rezero_weights)
+
+        if self._use_automatic_entropy_tuning:
+            alpha_loss = self._temperature_objective(log_pi_new_actions,
+                                                     samples_data)
+            self._alpha_optimizer.zero_grad()
+            alpha_loss.backward()
+            self._alpha_optimizer.step()
+
+        return policy_loss, qf1_loss, qf2_loss
+
+    def _evaluate_policy(self, epoch):
```

Comment:
Shouldn't this be the default behavior for all algorithms (to evaluate in eval mode), whether they are sparse or not? If that is the case, why it isn't implemented like this in garage?